security/advancedtls/provider.go

```diff
@@ -0,0 +1,199 @@
+		o.Interval = defaultInterval
+	}
+	ticker := time.NewTicker(o.Interval)
+	quit := make(chan bool)
+	provider := &IdentityPemFileProvider{}
+	provider.distributor = certprovider.NewDistributor()
+	// Initialize the distributor with an empty KeyMaterial.
+	provider.distributor.Set(&certprovider.KeyMaterial{}, nil)
+	// A goroutine to pull file changes.
+	go func() {
+		for {
+			select {
+			case <-quit:
+				return
+			case <-ticker.C:
+				{
+					// Read identity certs from PEM files.
+					identityCert, err := tls.LoadX509KeyPair(o.CertFile, o.KeyFile)
+					// If the reading produces an error, we will skip the update for this round and log the error.
+					// Note that LoadX509KeyPair will return error if file is empty,
+					// so there is no separate check for empty file contents.
+					if err != nil {
+						logger.Warning("tls.LoadX509KeyPair(%v, %v) failed: %v", o.CertFile, o.KeyFile, err)
+						continue
+					}
+					km := certprovider.KeyMaterial{Certs: []tls.Certificate{identityCert}}
+					provider.distributor.Set(&km, nil)
+				}
+			default:
+			}
+		}
+	}()
+	provider.closeFunc = func() {
+		ticker.Stop()
+		close(quit)
+	}
+	return provider, nil
+}
+
+// KeyMaterial returns the key material sourced by the IdentityPemFileProvider.
+// Callers are expected to use the returned value as read-only.
+func (p *IdentityPemFileProvider) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	return p.distributor.KeyMaterial(ctx)
+}
+
+// Close cleans up resources allocated by the IdentityPemFileProvider.
+func (p *IdentityPemFileProvider) Close() {
+	p.closeFunc()
+	p.distributor.Stop()
+}
+
+// NewRootPemFileProvider uses RootPemFileProviderOptions to construct a RootPemFileProvider.
+func NewRootPemFileProvider(o *RootPemFileProviderOptions) (*RootPemFileProvider, error) {
+	if o.TrustFile == "" {
+		return nil, fmt.Errorf("users need to specify key file path in RootPemFileProviderOptions")
+	}
+	// If interval is not set by users explicitly, we will set it to default interval.
+	if o.Interval == 0*time.Nanosecond {
+		o.Interval = defaultInterval
+	}
+	ticker := time.NewTicker(o.Interval)
+	quit := make(chan bool)
+	provider := &RootPemFileProvider{}
+	provider.distributor = certprovider.NewDistributor()
+	// Initialize the distributor with an empty KeyMaterial.
+	provider.distributor.Set(&certprovider.KeyMaterial{}, nil)
+	// A goroutine to pull file changes.
+	go func() {
+		for {
+			select {
+			case <-quit:
+				return
+			case <-ticker.C:
+				{
+					// If the current file is empty, skip the update for this round.
+					if trustFileSize, _ := getFileSize(o.TrustFile); trustFileSize == 0 {
+						continue
+					}
+					// Read root certs from PEM files.
+					trustData, err := ioutil.ReadFile(o.TrustFile)
+					trustPool := x509.NewCertPool()
+					trustPool.AppendCertsFromPEM(trustData)
+					// If the reading produces an error, we will skip the update for this round and log the error.
+					if err != nil {
```

Comment:
Please handle the error upfront and do a `continue`.

Reply:
Sure